zebra-test/src/transcript.rs

```diff
@@ -1,59 +1,144 @@
+                    Ok(expected_rsp) => Err(eyre!(e.into()))
+                        .wrap_err("service's ready_and returned an error: transcript expected an response")
+                        .with_section(|| {
+                            format!("{:?}", expected_rsp).header("Expected Response:")
+                        }),
+                    Err(error_checker) => {
+                        error_checker(e.into())
+                            .map_err(|e| eyre!(e))
+                            .wrap_err("service returned an error but it didn't match the expected error")?;
+                        continue;
+                    },
+                },
+            }?;
+
+            let response = fut.call(req).await;
+
+            match (response, expected_rsp) {
+                (Ok(rsp), Ok(expected_rsp)) => {
+                    if rsp != expected_rsp {
+                        Err(eyre!(
+                            "response doesn't match transcript's expected response"
+                        ))
+                        .with_section(|| format!("{:?}", expected_rsp).header("Expected Response:"))
+                        .with_section(|| format!("{:?}", rsp).header("Found Response:"))?;
+                    }
+                }
+                (Ok(rsp), Err(_)) => {
+                    todo!("got response when an error was expected. rsp={:?}", rsp)
+                }
+                (Err(_), Ok(expected_rsp)) => todo!(
+                    "got an error when a response was expected. expected_rsp={:?}",
+                    expected_rsp
+                ),
+                (Err(e), Err(error_checker)) => {
+                    error_checker(e.into()).map_err(|e| eyre!(e)).wrap_err(
+                        "service returned an error but it didn't match the expected error",
+                    )?;
+                    continue;
+                }
+            }
         }
         Ok(())
     }
 }
 
-impl<R, S, I> Service<R> for Transcript<R, S, I>
+const TRANSCRIPT_MOCK_GUIDE: &str = r#"
+Transcripts that are mocking services that produce errors must construct the
+errors they expect in their error handling closure.
+
+When a transcript encounters an expected error it passes in a `MockError`
+which the `check_fn` should then map to the desired error type.
+
+# Example
+
+```rust
+static EXAMPLE_TRANSCRIPT: Lazy<Vec<(Request, Response)>> = Lazy::new(|| {
+    let block = zebra_test::vectors::BLOCK_MAINNET_415000_BYTES
+        .zcash_deserialize_into::<Arc<_>>()
+        .unwrap();
+
+    vec![
+        (
+            Request::AddBlock {
+                block: block.clone(),
+            },
+            Err(|e| {
+                if e.is::<MockError>() {
+                    return Err(zebra_state::Error::ActualError);
+                } else if !e.is::<zebra_state::Error>() {
+                    todo!()
+                } else {
+                    Ok(())
+                }
+            })
```

Comment:
I'm not sure I understand this part.  Is this supposed to handle the "dual intent" usecase, where a `Transcript` is supposed to be able to be both a `Service` impl itself as well as to check an existing `Service`? Having to write this every time seems a little bit unergonomic but I'm not yet sure what a better alternative is.

Reply:
Yea, I was drafting a comment about how the `MockError` stuff was bad and asking for help on how to fix it when I realized I should just use `Option<BoxError>` as the input instead.